# Assignment 01
### Group 6 --  Yixuan Huang, Ada Li, Xinyu Wang, Jingsi Wu

### 1. Download files and decompress
https://www.kaggle.com/c/acquire-valued-shoppers-challenge/data

### 2. Read files into PySpark
You should now have 4 data sets:

* offers, a data set of different types of coupons – 6 columns, 37 rows
* testHistory, one shopper per row, used to test a predictive model – 5 columns, 151484 rows
* trainHistory, one shopper per row, used to train a predictive model – 7 columns, 160057 rows
* transactions, a detailed data set of shopper’s transactions – 11 columns, 349655789 rows

In [20]:
# Uncomment below if needed

# from pyspark.sql import SparkSession
# spark=SparkSession \
#         .builder \
#         .appName('assignment1') \
#         .getOrCreate()
# from pyspark.sql import functions

In [1]:
## modify path accordingly
path = '../../downloads/acquire-valued-shoppers-challenge/'
transactions = spark.read.option('header', 'true').csv(path + 'transactions.csv')
train = spark.read.option('header', 'true').csv(path + 'trainHistory.csv')
test = spark.read.option('header', 'true').csv(path + 'testHistory.csv')
offers = spark.read.option('header', 'true').csv(path + 'offers.csv')

In [2]:
# check data set
transactions.show()
train.show()
test.show()
offers.show()

+-----+-----+----+--------+----------+-----+----------+-----------+--------------+----------------+--------------+
|   id|chain|dept|category|   company|brand|      date|productsize|productmeasure|purchasequantity|purchaseamount|
+-----+-----+----+--------+----------+-----+----------+-----------+--------------+----------------+--------------+
|86246|  205|   7|     707|1078778070|12564|2012-03-02|         12|            OZ|               1|          7.59|
|86246|  205|  63|    6319| 107654575|17876|2012-03-02|         64|            OZ|               1|          1.59|
|86246|  205|  97|    9753|1022027929|    0|2012-03-02|          1|            CT|               1|          5.99|
|86246|  205|  25|    2509| 107996777|31373|2012-03-02|         16|            OZ|               1|          1.99|
|86246|  205|  55|    5555| 107684070|32094|2012-03-02|         16|            OZ|               2|         10.38|
|86246|  205|  97|    9753|1021015020|    0|2012-03-02|          1|            C

In [3]:
# show schema
transactions.printSchema()
train.printSchema()
test.printSchema()
offers.printSchema()

root
 |-- id: string (nullable = true)
 |-- chain: string (nullable = true)
 |-- dept: string (nullable = true)
 |-- category: string (nullable = true)
 |-- company: string (nullable = true)
 |-- brand: string (nullable = true)
 |-- date: string (nullable = true)
 |-- productsize: string (nullable = true)
 |-- productmeasure: string (nullable = true)
 |-- purchasequantity: string (nullable = true)
 |-- purchaseamount: string (nullable = true)

root
 |-- id: string (nullable = true)
 |-- chain: string (nullable = true)
 |-- offer: string (nullable = true)
 |-- market: string (nullable = true)
 |-- repeattrips: string (nullable = true)
 |-- repeater: string (nullable = true)
 |-- offerdate: string (nullable = true)

root
 |-- id: string (nullable = true)
 |-- chain: string (nullable = true)
 |-- offer: string (nullable = true)
 |-- market: string (nullable = true)
 |-- offerdate: string (nullable = true)

root
 |-- offer: string (nullable = true)
 |-- category: string (nullable = true)
 

In [4]:
# change variable type in transaction
transactions = transactions.withColumn("id", transactions.id.cast('long'))
transactions = transactions.withColumn("chain", transactions.chain.cast('int'))
transactions = transactions.withColumn("dept", transactions.dept.cast('int'))
transactions = transactions.withColumn("category", transactions.category.cast('int'))
transactions = transactions.withColumn("company", transactions.company.cast('int'))
transactions = transactions.withColumn("brand", transactions.brand.cast('int'))
transactions = transactions.withColumn("date", transactions.date.cast('timestamp'))
transactions = transactions.withColumn("productsize", transactions.productsize.cast('double'))
transactions = transactions.withColumn("purchaseamount", transactions.purchaseamount.cast('double'))
transactions = transactions.withColumn("purchasequantity", transactions.purchasequantity.cast('double'))
transactions.printSchema()

root
 |-- id: long (nullable = true)
 |-- chain: integer (nullable = true)
 |-- dept: integer (nullable = true)
 |-- category: integer (nullable = true)
 |-- company: integer (nullable = true)
 |-- brand: integer (nullable = true)
 |-- date: timestamp (nullable = true)
 |-- productsize: double (nullable = true)
 |-- productmeasure: string (nullable = true)
 |-- purchasequantity: double (nullable = true)
 |-- purchaseamount: double (nullable = true)



In [5]:
transactions.count()

349655789

In [5]:
# change variable type in train
train = train.withColumn("id", train.id.cast('long'))
train = train.withColumn("chain", train.chain.cast('int'))
train = train.withColumn("offer", train.offer.cast('int'))
train = train.withColumn("market", train.market.cast('int'))
train = train.withColumn("offerdate", train.offerdate.cast('timestamp'))
train.printSchema()

root
 |-- id: long (nullable = true)
 |-- chain: integer (nullable = true)
 |-- offer: integer (nullable = true)
 |-- market: integer (nullable = true)
 |-- repeattrips: string (nullable = true)
 |-- repeater: string (nullable = true)
 |-- offerdate: timestamp (nullable = true)



In [7]:
train.count()

160057

In [6]:
# change variable type in test
test = test.withColumn("id", test.id.cast('long'))
test = test.withColumn("chain", test.chain.cast('int'))
test = test.withColumn("offer", test.offer.cast('int'))
test = test.withColumn("market", test.market.cast('int'))
test = test.withColumn("offerdate", test.offerdate.cast('timestamp'))
test.printSchema()

root
 |-- id: long (nullable = true)
 |-- chain: integer (nullable = true)
 |-- offer: integer (nullable = true)
 |-- market: integer (nullable = true)
 |-- offerdate: timestamp (nullable = true)



In [9]:
test.count()

151484

In [7]:
# change variable type in offers
offers = offers.withColumn("offer", offers.offer.cast('int'))
offers = offers.withColumn("category", offers.category.cast('int'))
offers = offers.withColumn("quantity", offers.quantity.cast('int'))
offers = offers.withColumn("company", offers.company.cast('int'))
offers = offers.withColumn("offervalue", offers.offervalue.cast('double'))
offers = offers.withColumn("brand", offers.brand.cast('int'))
offers.printSchema()

root
 |-- offer: integer (nullable = true)
 |-- category: integer (nullable = true)
 |-- quantity: integer (nullable = true)
 |-- company: integer (nullable = true)
 |-- offervalue: double (nullable = true)
 |-- brand: integer (nullable = true)



In [11]:
offers.count()

37

### 3. The target variable for the contest was repeater. Remove repeattrips from the trainHistory set.
After dropping the repeattrips variable, the trainHistory set should now have 6 columns and 160057 rows.

In [8]:
# remove repeattrips
train = train.drop('repeattrips')
train.printSchema()

root
 |-- id: long (nullable = true)
 |-- chain: integer (nullable = true)
 |-- offer: integer (nullable = true)
 |-- market: integer (nullable = true)
 |-- repeater: string (nullable = true)
 |-- offerdate: timestamp (nullable = true)



In [13]:
train.count()

160057

### 4. Join the offers set onto the trainHistory and testHistory sets.
The new training set should contain 11 columns and 160057 rows. 

The new test set should contain 10 columns and 151484 rows.

In [9]:
# join offer on train history
joined_train = train.join(offers, 'offer', 'left')
joined_train.printSchema()

root
 |-- offer: integer (nullable = true)
 |-- id: long (nullable = true)
 |-- chain: integer (nullable = true)
 |-- market: integer (nullable = true)
 |-- repeater: string (nullable = true)
 |-- offerdate: timestamp (nullable = true)
 |-- category: integer (nullable = true)
 |-- quantity: integer (nullable = true)
 |-- company: integer (nullable = true)
 |-- offervalue: double (nullable = true)
 |-- brand: integer (nullable = true)



In [15]:
joined_train.count()

160057

In [10]:
# join offer on test history
joined_test = test.join(offers, 'offer', 'left')
joined_test.printSchema()

root
 |-- offer: integer (nullable = true)
 |-- id: long (nullable = true)
 |-- chain: integer (nullable = true)
 |-- market: integer (nullable = true)
 |-- offerdate: timestamp (nullable = true)
 |-- category: integer (nullable = true)
 |-- quantity: integer (nullable = true)
 |-- company: integer (nullable = true)
 |-- offervalue: double (nullable = true)
 |-- brand: integer (nullable = true)



In [17]:
joined_test.count()

151484

### 5. Determine the average number of items and the average dollar amount a shopper has spent on items in the same category as the item for which they received an offer in the trainHistory or testHistory sets.
After the join operation, the new training set should contain 13 columns and 160057 rows. 

The new test set should now contain 12 columns and 151484 rows.

In [11]:
# group transaction by id and category, and calculate average quant and amount
add = transactions.groupby('id', 'category').avg('purchasequantity','purchaseamount')
add = add.withColumnRenamed('avg(purchasequantity)', 'mean_quantity')
add = add.withColumnRenamed('avg(purchaseamount)', 'mean_amount')
add.show()

+--------+--------+------------------+------------------+
|      id|category|     mean_quantity|       mean_amount|
+--------+--------+------------------+------------------+
|   86246|    2116|1.3333333333333333|               1.5|
|   86246|    3108|1.0357142857142858|1.3692857142857142|
|   86252|     905|              1.52|            2.6744|
|   86252|     916|               1.0|              3.79|
|   86252|    9781|               1.0|              4.44|
|12262064|    6330|               1.0|              3.69|
|12262064|    2301|              1.25|              1.82|
|12277270|    2709|               2.0|              2.78|
|12524696|    5823|1.1111111111111112| 5.224444444444445|
|12524696|     416|2.2222222222222223|1.8233333333333333|
|12524696|    4401|               1.0|              4.25|
|12682470|     421|               1.0|               2.5|
|12682470|    3204|               1.0|              9.29|
|13074629|    9716|1.3333333333333333|             13.32|
|13251776|    

In [12]:
# join add on test history
joined_test = joined_test.join(add, ['id', 'category'], 'left').fillna(0)
joined_test.printSchema()

root
 |-- id: long (nullable = true)
 |-- category: integer (nullable = true)
 |-- offer: integer (nullable = true)
 |-- chain: integer (nullable = true)
 |-- market: integer (nullable = true)
 |-- offerdate: timestamp (nullable = true)
 |-- quantity: integer (nullable = true)
 |-- company: integer (nullable = true)
 |-- offervalue: double (nullable = false)
 |-- brand: integer (nullable = true)
 |-- mean_quantity: double (nullable = false)
 |-- mean_amount: double (nullable = false)



In [20]:
joined_test.count()

151484

In [13]:
# join add on train history
joined_train = joined_train.join(add, ['id', 'category'], 'left').fillna(0)
joined_train.printSchema()

root
 |-- id: long (nullable = true)
 |-- category: integer (nullable = true)
 |-- offer: integer (nullable = true)
 |-- chain: integer (nullable = true)
 |-- market: integer (nullable = true)
 |-- repeater: string (nullable = true)
 |-- offerdate: timestamp (nullable = true)
 |-- quantity: integer (nullable = true)
 |-- company: integer (nullable = true)
 |-- offervalue: double (nullable = false)
 |-- brand: integer (nullable = true)
 |-- mean_quantity: double (nullable = false)
 |-- mean_amount: double (nullable = false)



In [23]:
joined_train.count()

160057

### 6. Determine if the customer has previously purchased the exact item on offer.
After the join operation, the new training set should contain 14 columns and 160057 rows.

The new test set should now contain 13 columns and 151484 rows.

### Assumptions:
Our team defines that once a customer bought this item whether returns it or not, the customer is considered previously purchased the exact item on offer.

Under this circumstance, our result shows that 25215 customers bought the item before offer in train data set.

In [14]:
# group the transaction by customer and the items feature, find the max purchased quantity.
add2 = transactions.groupby('id', 'category', 'company', 'brand').max('purchasequantity')
add2 = add2.withColumnRenamed('max(purchasequantity)', 'purchased')
# filter out the records with no purchased quantity
add2 = add2.filter(add2.purchased > 0)
add2.show()

+-----+--------+----------+-----+---------+
|   id|category|   company|brand|purchased|
+-----+--------+----------+-----+---------+
|86246|    3410| 107418272|16853|      1.0|
|86246|    2105|1065762262|26129|      1.0|
|86246|    6408| 108072080| 7265|      1.0|
|86246|    2620| 104113040|28204|      2.0|
|86246|    2806|1078099373| 6126|      1.0|
|86246|    8101| 102820020|19426|      1.0|
|86246|    2629| 104130343|38922|      1.0|
|86246|    6324| 103507131|54544|      1.0|
|86246|    1827|1080882989|10794|      1.0|
|86246|    3202| 101200515|71182|      1.0|
|86246|     921| 107247070| 9903|      2.0|
|86246|     908| 102951929| 4628|      1.0|
|86246|    6409|1071582676|14951|      1.0|
|86246|    2702| 107910070| 7608|      1.0|
|86246|    2112| 104473848|21039|      1.0|
|86246|    4108| 104144343|11130|      2.0|
|86252|    6011| 101800010|14029|      3.0|
|86252|    1898| 103980030| 5782|      1.0|
|86252|     970| 103800030|62065|      2.0|
|86252|    3301| 104178040|19154

In [15]:
# change purchased records into boolean values(1 for purchased, 0 for not)
# in this case, all records in add2 should be 1 at purchased column
add2 = add2.withColumn('purchased', (add2.purchased > 0))
add2 = add2.withColumn('purchased', add2.purchased.cast('int'))
add2.show()

+-----+--------+----------+-----+---------+
|   id|category|   company|brand|purchased|
+-----+--------+----------+-----+---------+
|86246|    3410| 107418272|16853|        1|
|86246|    2105|1065762262|26129|        1|
|86246|    6408| 108072080| 7265|        1|
|86246|    2620| 104113040|28204|        1|
|86246|    2806|1078099373| 6126|        1|
|86246|    8101| 102820020|19426|        1|
|86246|    2629| 104130343|38922|        1|
|86246|    6324| 103507131|54544|        1|
|86246|    1827|1080882989|10794|        1|
|86246|    3202| 101200515|71182|        1|
|86246|     921| 107247070| 9903|        1|
|86246|     908| 102951929| 4628|        1|
|86246|    6409|1071582676|14951|        1|
|86246|    2702| 107910070| 7608|        1|
|86246|    2112| 104473848|21039|        1|
|86246|    4108| 104144343|11130|        1|
|86252|    6011| 101800010|14029|        1|
|86252|    1898| 103980030| 5782|        1|
|86252|     970| 103800030|62065|        1|
|86252|    3301| 104178040|19154

In [16]:
# join add2 on test and train set
joined_test = joined_test.join(add2, ['id', 'category', 'company', 'brand'], 'left')
joined_train = joined_train.join(add2, ['id', 'category', 'company', 'brand'], 'left')

In [17]:
# fill null with 0, and check schema
joined_test = joined_test.fillna(0)
joined_train = joined_train.fillna(0)
joined_test.printSchema()
joined_train.printSchema()

root
 |-- id: long (nullable = true)
 |-- category: integer (nullable = true)
 |-- company: integer (nullable = true)
 |-- brand: integer (nullable = true)
 |-- offer: integer (nullable = true)
 |-- chain: integer (nullable = true)
 |-- market: integer (nullable = true)
 |-- offerdate: timestamp (nullable = true)
 |-- quantity: integer (nullable = true)
 |-- offervalue: double (nullable = false)
 |-- mean_quantity: double (nullable = false)
 |-- mean_amount: double (nullable = false)
 |-- purchased: integer (nullable = true)

root
 |-- id: long (nullable = true)
 |-- category: integer (nullable = true)
 |-- company: integer (nullable = true)
 |-- brand: integer (nullable = true)
 |-- offer: integer (nullable = true)
 |-- chain: integer (nullable = true)
 |-- market: integer (nullable = true)
 |-- repeater: string (nullable = true)
 |-- offerdate: timestamp (nullable = true)
 |-- quantity: integer (nullable = true)
 |-- offervalue: double (nullable = false)
 |-- mean_quantity: double (n

In [19]:
# number of customers bought the item before offer
joined_train.filter(joined_train.purchased == 1).count()

25215

### 7. Create a small subset of the training data, export it to CSV, and turn it in.
Subset the augmented training data so that it contains only id numbers less than 107014607 where average purchase quantity and average purchase amount are greater than zero. 

Export this small file to CSV format.

In [18]:
# subset train set with the given requirement
joined_train_subset = joined_train.filter((joined_train.id < 107014607) \
                                          & (joined_train['mean_amount'] > 0) \
                                          & (joined_train['mean_quantity'] > 0))

In [22]:
joined_train_subset.printSchema()

root
 |-- id: long (nullable = true)
 |-- category: integer (nullable = true)
 |-- company: integer (nullable = true)
 |-- brand: integer (nullable = true)
 |-- offer: integer (nullable = true)
 |-- chain: integer (nullable = true)
 |-- market: integer (nullable = true)
 |-- repeater: string (nullable = true)
 |-- offerdate: timestamp (nullable = true)
 |-- quantity: integer (nullable = true)
 |-- offervalue: double (nullable = false)
 |-- mean_quantity: double (nullable = false)
 |-- mean_amount: double (nullable = false)
 |-- purchased: integer (nullable = true)



In [19]:
# write into csv
joined_train_subset.toPandas().to_csv('train_subset.csv')

In [20]:
# check output
import pandas as pd
output = pd.read_csv('train_subset.csv')
output.shape

(350, 15)

We can see that the final data set contains 350 rows，14 columns + 1 column of index.

In [21]:
output.head()

Unnamed: 0         id  category    company  brand    offer  chain  market  \
0           0   99128004      5616  104610040  15889  1204576     17       4   
1           1  100470893      5616  104610040  15889  1204576     15       9   
2           2   97167859      3203  106414464  13474  1197502      4       1   
3           3   98319236      1726  104460040   7668  1200581     17       4   
4           4   99780431      3203  106414464  13474  1197502     15       9   

  repeater   offerdate  quantity  offervalue  mean_quantity  mean_amount  \
0        f  2013-04-06         1        1.00            1.5     3.417500   
1        f  2013-04-07         1        1.00            1.0     4.656667   
2        t  2013-04-02         1        0.75            1.0     3.590000   
3        f  2013-03-30         1        1.25            1.0     2.374000   
4        f  2013-03-27         1        0.75            2.0     8.980000   

   purchased  
0          1  
1          0  
2          0  
3          0  
4          0